test/Dialect/FIRRTL/canonicalization.mlir

```diff
@@ -1249,4 +1249,21 @@ firrtl.module @mul_cst_prop3(%out_b: !firrtl.flip<sint<15>>) {
   firrtl.connect %out_b, %add : !firrtl.flip<sint<15>>, !firrtl.sint<15>
 }
 
+// We fold `validif` operations to their RHS, regardless of the LHS.
+// See https://github.com/llvm/circt/issues/839.
+// CHECK-LABEL: @elide_validif
+// CHECK-NEXT:      firrtl.connect %out0, %clock : !firrtl.flip<clock>, !firrtl.clock
+// CHECK-NEXT:      firrtl.connect %out1, %clock : !firrtl.flip<clock>, !firrtl.clock
+// CHECK-NEXT:  }
+firrtl.module @elide_validif(%clock: !firrtl.clock, %valid: !firrtl.uint<1>, %out0: !firrtl.flip<clock>, %out1: !firrtl.flip<clock>) {
+  %x = firrtl.wire  : !firrtl.clock
+  %y = firrtl.wire  : !firrtl.clock
+  %0 = firrtl.validif %valid, %clock : (!firrtl.uint<1>, !firrtl.clock) -> !firrtl.clock
+  %1 = firrtl.validif %valid, %clock : (!firrtl.uint<1>, !firrtl.clock) -> !firrtl.clock
+  firrtl.connect %x, %0 : !firrtl.clock, !firrtl.clock
+  firrtl.connect %y, %1 : !firrtl.clock, !firrtl.clock
+  firrtl.connect %out0, %x : !firrtl.flip<clock>, !firrtl.clock
+  firrtl.connect %out1, %y : !firrtl.flip<clock>, !firrtl.clock
+}
+
```

Comment:
It looks like this is only checking inputs of type clock, should we also check a UInt or something?

Reply:
We can do that. The fold is invariant to the input type though, so basically any input type will be forwarded. If it works for `!firrtl.clock` the MLIR infrastructure should guarantee that it works for all the other types as well.